### PR TITLE
Merchant Traders now have greetings for Species + Biases

### PR DESCRIPTION
**Pull request by Ben10083**
Traders used to only do greetings based on each subspecies, due to significant overlap for some (ie, IPCs), functionality added for traders to greet based on general species, if a specific subspecies is not found.

Adds new defines for each species that is a list that has all subspecies for them (ALL_IPC covers all frames for examples)

Biases for Species
DENY: "No service for Bots here"
UPCHARGE: Service, but prices higher
DISCOUNT: Special prrrices forrr comrrrades

### CODE REVIEW

**Comment by Ben10083** [comment]
Going to look into adding racism to the traders (this isnt a joke)

**Comment by Ben10083** [comment]
!wip

File: code/__DEFINES/species.dm [hunk 1]
```diff
@@ -112,3 +112,31 @@
...
+
+#define ALL_VAURCA list(\
+		SPECIES_VAURCA_WORKER, SPECIES_VAURCA_WARRIOR, SPECIES_VAURCA_BREEDER, SPECIES_VAURCA_WARFORM, SPECIES_VAURCA_BULWARK, \
+	)
+
+#define ALL_IPCS list(\
+		SPECIES_IPC, SPECIES_IPC_BISHOP, SPECIES_IPC_G1, SPECIES_IPC_G2, SPECIES_IPC_SHELL, \
+		SPECIES_IPC_UNBRANDED, SPECIES_IPC_UNBRANDED_REMOTE, SPECIES_IPC_XION, SPECIES_IPC_ZENGHU, \
+		SPECIES_IPC_SHELL_ROGUE, SPECIES_IPC_XION_REMOTE, SPECIES_IPC_PURPOSE_HK, \
+	)
```
Comment (DreamySkrell) [inline code/__DEFINES/species.dm:140]: All of these should have names like `ALL_HUMAN_SPECIES` to be a bit more consistent with other defines

**Comment by Ben10083** [comment]
!review

**Comment by Ben10083** [comment]
It is done.

File: code/datums/trading/trade.dm [hunk 2]
```diff
@@ -73,6 +95,30 @@
...
+	if(!species || !species_bias) // No species to be biased against
+		return FALSE
+
+	// Flatten list
+	for(var/item in species_bias)
+		var/result = species_bias[item]
+		if(islist(item))
+			for(var/flat_item in flatten_list(item))
+				if(species == flat_item)
+					return result
```
Comment (NonQueueingMatt) [inline code/datums/trading/trade.dm:116]:
```
if(species in flatten_list(item))
    return result
```

File: code/datums/trading/trade.dm [hunk 3]
```diff
@@ -175,21 +229,45 @@
...
+			if(H.species.name in ALL_DIONA_SPECIES)
+				general = "Diona"
+			if(H.species.name in ALL_SKRELL_SPECIES)
+				general = "Skrell"
+			if(H.species.name in ALL_TAJARA_SPECIES)
+				general = "Tajara"
+			if(H.species.name in ALL_VAURCA_SPECIES)
+				general = "Vaurca"
+			if(H.species.name in ALL_IPC_SPECIES)
+				general = "IPC"
```
Comment (NonQueueingMatt) [inline code/datums/trading/trade.dm:251]: use a keyed list here instead of a massive if chain

**Review by NonQueueingMatt — CHANGES_REQUESTED** [review]

File: code/datums/trading/weaponry.dm [hunk 4]
```diff
@@ -100,6 +101,12 @@
 		"insult_good"       = "Good one, comrrrade!",
 		"insult_bad"        = "Rrrracist!"
 	)
+	species_bias = list( //Zhan don't get discount
+		SPECIES_TAJARA = TRADER_BIAS_DISCOUNT,
+		SPECIES_TAJARA_ZHAN = TRADER_BIAS_DISCOUNT,
```
Comment (Generalcamo) [inline code/datums/trading/weaponry.dm:106]: You're contradicting the comment above.

**Review by Generalcamo — CHANGES_REQUESTED** [review]

**Review by Generalcamo — APPROVED** [review]

**Review by Alberyk — APPROVED** [review]

**Review by DreamySkrell — APPROVED** [review]

**Pull request merged**